sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RequestTimeline.java

```diff
@@ -0,0 +1,181 @@
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.databind.ser.std.ToStringSerializer;
+import com.google.common.collect.ImmutableList;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.Iterator;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Represents the time and duration of important events in the lifetime of a request.
+ *
+ * A {@link RequestTimeline} represents a timeline as a sequence of {@link Event} instances with name, time, and
+ * duration properties. Hence, one might use this class to represent any timeline. Today we use it to represent
+ * {@link RntbdTransportClient} request timelines. In the future we might also use it to represent
+ * {@link HttpTransportClient} request timelines.
+ *
+ * A {@link RequestTimeline} serializes to JSON as an array of {@link Event} instances. This is the default
+ * serialization for any class that implements {@link Iterable}.
+ * <p>
+ * <b>Example:</b>
+ * <pre>{@code OffsetDateTime startTime = OffsetDateTime.parse("2020-01-07T11:24:12.842749-08:00", DateTimeFormatter.ISO_OFFSET_DATE_TIME);
+ * sys.out.println(RequestTimeline.of(
+ *     new RequestTimeline.Event("foo", startTime, startTime.plusSeconds(1)),
+ *     new RequestTimeline.Event("bar", startTime.plusSeconds(1), startTime.plusSeconds(2))));}</pre>
+ * JSON serialization:
+ * <pre>{@code [{"name":"foo","time":"2020-01-07T11:24:12.842749-08:00","duration":"PT1S"},{"name":"bar","time":"2020-01-07T11:24:13.842749-08:00","duration":"PT1S"}])}</pre>
+ */
+public final class RequestTimeline implements Iterable<RequestTimeline.Event> {
+
+    private static final RequestTimeline EMPTY = new RequestTimeline();
+    private final ImmutableList<Event> events;
+
+    private RequestTimeline() {
+        this.events = ImmutableList.of();
+    }
+
+    private RequestTimeline(final ImmutableList<Event> events) {
+        checkNotNull(events, "expected non-null events");
+        this.events = events;
+    }
+
+    /**
+     * Returns an empty {@link RequestTimeline}.
+     *
+     * The empty time line returned is static.
+     *
+     * @return an empty {@link RequestTimeline}.
+     */
+    public static RequestTimeline empty() {
+        return EMPTY;
+    }
+
+    /**
+     * Returns an iterator for enumerating the {@link Event} instances in this {@link RequestTimeline}.
+     *
+     * @return an iterator for enumerating the {@link Event} instances in this {@link RequestTimeline}.
+     */
+    @Override
+    public Iterator<Event> iterator() {
+        return this.events.iterator();
+    }
+
+    /**
+     * Returns an empty {@link RequestTimeline}.
+     *
+     * The empty time line returned is static and equivalent to calling {@link RequestTimeline#empty}.
+     *
+     * @return an empty request timeline.
+     */
+    public static RequestTimeline of() {
```

Comment:
Any reason we need this in addition to `empty()` API above ?
Is it okay to share the `EMPTY` static instance. If multiple callers get this object and modify it, the `EMPTY` static instance will be modified, right ?
Or if it is singleton, then looks good.

Reply:
EMPTY is a singleton and RequestTimeline instances are immutable. The of() method is provided for completeness and could easily be removed. I thought about removing it once or twice myself. I'll resolve this issue assuming that it's OK to leave the of() method in.